Revert https://github.com/llvm/llvm-project/commit/ea413cbacf7f268a94e980958c970cb3a20cfc4d

Upstream emulates "ld main.o x.a y.a x.so" by using x symbols needed for y.a from x.so,
whereas we want to emulate "ld main.o --start-group x.a y.a --end-group x.so" and use x.a.

Issue: https://st.yandex-team.ru/DTCC-167

--- a/ELF/Symbols.cpp
+++ b/ELF/Symbols.cpp
@@ -746 +746 @@ void Symbol::resolveShared(const SharedSymbol &other) {
-  if (visibility == STV_DEFAULT && (isUndefined() || isLazy())) {
+  if (visibility == STV_DEFAULT && isUndefined()) {
